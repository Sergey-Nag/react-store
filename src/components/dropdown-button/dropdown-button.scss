.dropdown-right {
  right: 0;
  top: 100%;
}