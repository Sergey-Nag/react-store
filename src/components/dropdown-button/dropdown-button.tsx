import React, { Children, cloneElement, ReactElement, ReactNode, useEffect, useRef, useState } from "react";
import { ChevronUp, ChevronDown } from "react-bootstrap-icons";
import './dropdown-button.scss';
import cn from 'classnames';

interface DropdownButtonProps {
  text?: string | ReactElement;
  children: ReactNode;
  [key: string]: any;
}

export function DropdownButton({ text, className, children }: DropdownButtonProps) {
  const buttonClassNames = cn(className, 'btn');
  const buttonRef = useRef<HTMLDivElement>(null);
  const [isDropdownShown, setIsDropdownShown] = useState(false);
  const dropdownClassNames = cn('dropdown-menu', 'dropdown-right', isDropdownShown && 'show');

  const modifyChildren = (child: ReactNode) => {
    if (child && 'props' in (child as ReactElement)) {
      const className = cn(
        (child as ReactElement).props.className,
        'dropdown-item'
      );

      return cloneElement(child as ReactElement, { className, role: 'menuitem' });
    }

    return child;
  }
  
  useEffect(() => {
    const handleOutsideClick = ({ target }: MouseEvent) => {
      if (target && buttonRef.current?.contains(target as HTMLElement)) {
        setIsDropdownShown((isShown) => !isShown);

        return;
      };

      setIsDropdownShown(false);
    }

    document.addEventListener('click', handleOutsideClick);

    return () => document.removeEventListener('click', handleOutsideClick);
  }, [buttonRef]);

  return (
    <div
      className="position-relative"
      role="menu"
      ref={buttonRef}
    >
      <button type="button" className={buttonClassNames}>
        { text || (isDropdownShown ? <ChevronUp /> : <ChevronDown />) }
      </button>
      <div className={dropdownClassNames}>
        { Children.map(children, modifyChildren) }
      </div>
    </div>
  );
}
