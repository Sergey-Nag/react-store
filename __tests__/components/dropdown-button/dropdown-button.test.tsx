import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import { DropdownButton, DropdownButtonProps } from "../../../src/components/dropdown-button/dropdown-button";

const renderWithItems = (rest?: DropdownButtonProps) => render(
  <DropdownButton {...rest}>
    <div>Test item 1</div>
    <div>Test item 2</div>
    Test text 3
  </DropdownButton>
);

describe('Dropdown button', () => {
  it('Should initially render component without nested items', () => {
    const { asFragment } = renderWithItems();

    expect(asFragment()).toMatchSnapshot();
  });

  it('Should render component with nested elements to which the menuitem class has been passed', () => {
    const { asFragment } = renderWithItems();

    fireEvent.click(screen.getByRole('button'));

    expect(asFragment()).toMatchSnapshot();
  });

  it('Should opend dropdown with nested elements when button is clicked and close after clicking outside', async () => {
    renderWithItems();

    expect(screen.queryByRole('menuitem')).not.toBeInTheDocument();
    
    fireEvent.click(screen.getByRole('button'));
    
    await waitFor(() => {
      expect(screen.getAllByRole('menuitem')).toHaveLength(2);
    });

    await waitFor(() => {
      expect(screen.getByText('Test text 3')).toBeInTheDocument();
    });

    fireEvent.click(document.body);

    await waitFor(() => {
      expect(screen.queryByRole('menuitem')).not.toBeInTheDocument();
    });
  });

  it('Should change button name if related prop passed', () => {
    renderWithItems({ text: 'Dropdown' });

    expect(screen.getByRole('button', { name: 'Dropdown' })).toBeInTheDocument();
  });
});