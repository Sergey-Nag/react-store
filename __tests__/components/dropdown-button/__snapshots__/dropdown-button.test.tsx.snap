// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown button Should initially render component without nested items 1`] = `
<DocumentFragment>
  <div
    class="position-relative"
    role="menu"
  >
    <button
      class="btn"
      type="button"
    >
      <svg
        fill="currentColor"
        height="1em"
        viewBox="0 0 16 16"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M1.646 4.646a.5.5 0 0 1 .708 0L8 10.293l5.646-5.647a.5.5 0 0 1 .708.708l-6 6a.5.5 0 0 1-.708 0l-6-6a.5.5 0 0 1 0-.708z"
          fill-rule="evenodd"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;

exports[`Dropdown button Should render component with nested elements to which the menuitem class has been passed 1`] = `
<DocumentFragment>
  <div
    class="position-relative"
    role="menu"
  >
    <button
      class="btn"
      type="button"
    >
      <svg
        fill="currentColor"
        height="1em"
        viewBox="0 0 16 16"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M7.646 4.646a.5.5 0 0 1 .708 0l6 6a.5.5 0 0 1-.708.708L8 5.707l-5.646 5.647a.5.5 0 0 1-.708-.708l6-6z"
          fill-rule="evenodd"
        />
      </svg>
    </button>
    <div
      class="dropdown-menu dropdown-right show"
    >
      <div
        class="dropdown-item"
        role="menuitem"
      >
        Test item 1
      </div>
      <div
        class="dropdown-item"
        role="menuitem"
      >
        Test item 2
      </div>
      Test text 3
    </div>
  </div>
</DocumentFragment>
`;
